File: ui/src/api/hooks/simplefin.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import {
  commands,
  type SimpleFinAccountImportRequest,
  type SimpleFinStatus,
  type SimpleFinAccountInfo,
  type SimpleFinConnectionInfo,
  type SyncSummary,
  type SimpleFinSyncSettings,
  type SimpleFinAlert,
  type TransferSuggestionInfo,
  type ImportCandidateWithMatches,
} from "../client";
import { unwrap } from "../client";
import { invalidateDomains } from "../invalidation";

/**
 * Canonical keys for every simplefin query. Exported so screens that need to
 * invalidate one surface (e.g. the Inbox resolving alerts/transfers/import
 * review) reuse these instead of re-typing the shape by hand.
 */
export const simplefinKeys = {
  status: ["simplefin", "status"] as const,
  accounts: ["simplefin", "accounts"] as const,
  connections: ["simplefin", "connections"] as const,
  syncSettings: ["simplefin", "syncSettings"] as const,
  alerts: ["simplefin", "alerts"] as const,
  transfers: ["simplefin", "transfers"] as const,
  importReview: ["simplefin", "importReview"] as const,
};

export function useSimpleFinStatus() {
  return useQuery<SimpleFinStatus>({
    queryKey: simplefinKeys.status,
    queryFn: async () => {
      return unwrap(commands.getSimplefinStatus());
    },
  });
}

export function useSaveSimpleFinToken() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (token: string) => {
      return unwrap(commands.saveSimplefinSetupToken(token));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.status });
      qc.invalidateQueries({ queryKey: simplefinKeys.connections });
    },
  });
}

export function useSimpleFinConnections() {
  return useQuery<SimpleFinConnectionInfo[]>({
    queryKey: simplefinKeys.connections,
    queryFn: async () => {
      return unwrap(commands.listSimplefinConnections());
    },
  });
}

export function useSimpleFinAccounts() {
  return useQuery<SimpleFinAccountInfo[]>({
    queryKey: simplefinKeys.accounts,
    queryFn: async () => {
      return unwrap(commands.listSimplefinAccounts());
    },
    enabled: false,
  });
}

export function useImportSimpleFinAccounts() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (accounts: SimpleFinAccountImportRequest[]) => {
      return unwrap(commands.importSimplefinAccounts(accounts));
    },
    onSuccess: () => {
      // A committed SimpleFin import touches the whole ledger + accounts +
      // import state (previously under-invalidated month-totals/net-worth).
      invalidateDomains(qc, "simplefin");
      qc.invalidateQueries({ queryKey: simplefinKeys.accounts });
    },
  });
}

export function useSyncSimpleFinAccount() {
  const qc = useQueryClient();
  return useMutation<SyncSummary, Error, string>({
    mutationFn: async (accountId: string) => {
      return unwrap(commands.syncSimplefinAccount(accountId));
    },
    onSuccess: () => {
      // Sync adds/updates rows: full ledger fan-out, not just the two roots.
      invalidateDomains(qc, "simplefin");
    },
  });
}

export function useDisconnectSimpleFin() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async () => {
      return unwrap(commands.disconnectSimplefin());
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.status });
      qc.invalidateQueries({ queryKey: simplefinKeys.connections });
      invalidateDomains(qc, "accounts", "transactions");
    },
  });
}

export function usePurgeSimpleFinData() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async () => {
      return unwrap(commands.purgeSimplefinData());
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.status });
      qc.invalidateQueries({ queryKey: simplefinKeys.connections });
      qc.invalidateQueries({ queryKey: simplefinKeys.accounts });
      qc.invalidateQueries({ queryKey: simplefinKeys.alerts });
      qc.invalidateQueries({ queryKey: simplefinKeys.transfers });
      qc.invalidateQueries({ queryKey: simplefinKeys.importReview });
      invalidateDomains(qc, "simplefin");
      qc.invalidateQueries({ queryKey: ["onboarding"] });
    },
  });
}

export function useDeleteSimpleFinConnection() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (connectionId: string) => {
      return unwrap(commands.deleteSimplefinConnection(connectionId));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.connections });
      qc.invalidateQueries({ queryKey: simplefinKeys.status });
      invalidateDomains(qc, "accounts", "transactions");
    },
  });
}

export function useSyncAllSimpleFinAccounts() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async () => {
      return unwrap(commands.syncAllSimplefinAccounts());
    },
    onSuccess: () => {
      invalidateDomains(qc, "simplefin");
      qc.invalidateQueries({ queryKey: simplefinKeys.connections });
      qc.invalidateQueries({ queryKey: simplefinKeys.alerts });
      qc.invalidateQueries({ queryKey: simplefinKeys.importReview });
    },
  });
}

export function useSimpleFinSyncSettings() {
  return useQuery<SimpleFinSyncSettings>({
    queryKey: simplefinKeys.syncSettings,
    queryFn: async () => {
      return unwrap(commands.getSimplefinSyncSettings());
    },
  });
}

export function useSetSimpleFinSyncSettings() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (settings: SimpleFinSyncSettings) => {
      return unwrap(commands.setSimplefinSyncSettings(settings));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.syncSettings });
    },
  });
}

export function useSimpleFinAlerts() {
  return useQuery<SimpleFinAlert[]>({
    queryKey: simplefinKeys.alerts,
    queryFn: async () => {
      return unwrap(commands.listSimplefinAlerts());
    },
  });
}

export function useAcknowledgeSimpleFinAlert() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (alertId: string) => {
      return unwrap(commands.acknowledgeSimplefinAlert(alertId));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.alerts });
    },
  });
}

export function useSimpleFinTransferSuggestions() {
  return useQuery<TransferSuggestionInfo[]>({
    queryKey: simplefinKeys.transfers,
    queryFn: async () => {
      return unwrap(commands.listSimplefinTransferSuggestions());
    },
  });
}

export function useConfirmSimpleFinTransfer() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (transferId: string) => {
      return unwrap(commands.confirmSimplefinTransfer(transferId));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.transfers });
      invalidateDomains(qc, "transactions");
    },
  });
}

export function useRejectSimpleFinTransfer() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (transferId: string) => {
      return unwrap(commands.rejectSimplefinTransfer(transferId));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.transfers });
    },
  });
}

export function useImportReviewCandidates() {
  return useQuery<ImportCandidateWithMatches[]>({
    queryKey: simplefinKeys.importReview,
    queryFn: async () => {
      return unwrap(commands.listImportReviewCandidates());
    },
  });
}

export function useAcceptImportCandidateMatch() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ candidateId, transactionId }: { candidateId: string; transactionId: string }) => {
      return unwrap(commands.acceptImportCandidateMatch(candidateId, transactionId));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.importReview });
      invalidateDomains(qc, "simplefin");
    },
  });
}

export function useCreateImportCandidateTransaction() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (candidateId: string) => {
      return unwrap(commands.createImportCandidateTransaction(candidateId));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.importReview });
      invalidateDomains(qc, "simplefin");
    },
  });
}

export function useDismissImportCandidate() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (candidateId: string) => {
      return unwrap(commands.dismissImportCandidate(candidateId));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: simplefinKeys.importReview });
    },
  });
}

File: crates/finsight-bindings/src/commands/simplefin.rs
use crate::error::AppResult;
use crate::AppState;

// Types live in finsight-api now; re-exported so existing imports of
// `finsight_bindings::commands::simplefin::*` (lib.rs, tests) keep resolving.
pub use finsight_api::commands::simplefin::{
    SimpleFinAccountImportRequest, SimpleFinAccountInfo, SimpleFinConnectionInfo,
    SimpleFinPurgeSummary, SimpleFinStatus, SyncSummary, TransferSuggestionInfo,
};

use finsight_core::models::{ImportCandidateWithMatches, SimpleFinAlert};

/// Claim a SimpleFin setup token and persist the resulting bridge access URL
/// plus every connection exposed by that access URL.
#[tauri::command]
#[specta::specta]
pub async fn save_simplefin_setup_token(
    state: tauri::State<'_, AppState>,
    token: String,
) -> AppResult<Vec<SimpleFinConnectionInfo>> {
    finsight_api::commands::simplefin::save_simplefin_setup_token(&state.api, token).await
}

#[tauri::command]
#[specta::specta]
pub async fn get_simplefin_status(state: tauri::State<'_, AppState>) -> AppResult<SimpleFinStatus> {
    finsight_api::commands::simplefin::get_simplefin_status(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_simplefin_connections(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<SimpleFinConnectionInfo>> {
    finsight_api::commands::simplefin::list_simplefin_connections(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_simplefin_accounts(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<SimpleFinAccountInfo>> {
    finsight_api::commands::simplefin::list_simplefin_accounts(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn import_simplefin_accounts(
    state: tauri::State<'_, AppState>,
    accounts: Vec<SimpleFinAccountImportRequest>,
) -> AppResult<Vec<String>> {
    finsight_api::commands::simplefin::import_simplefin_accounts(&state.api, accounts).await
}

#[tauri::command]
#[specta::specta]
pub async fn sync_simplefin_account(
    state: tauri::State<'_, AppState>,
    account_id: String,
) -> AppResult<SyncSummary> {
    finsight_api::commands::simplefin::sync_simplefin_account(&state.api, account_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn disconnect_simplefin(state: tauri::State<'_, AppState>) -> AppResult<()> {
    finsight_api::commands::simplefin::disconnect_simplefin(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn purge_simplefin_data(
    state: tauri::State<'_, AppState>,
) -> AppResult<SimpleFinPurgeSummary> {
    finsight_api::commands::simplefin::purge_simplefin_data(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn delete_simplefin_connection(
    state: tauri::State<'_, AppState>,
    connection_id: String,
) -> AppResult<()> {
    finsight_api::commands::simplefin::delete_simplefin_connection(&state.api, connection_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn sync_all_simplefin_accounts(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<crate::sync_scheduler::AccountSyncResult>> {
    finsight_api::commands::simplefin::sync_all_simplefin_accounts(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn get_simplefin_sync_settings(
    state: tauri::State<'_, AppState>,
) -> AppResult<crate::sync_scheduler::SimpleFinSyncSettings> {
    finsight_api::commands::simplefin::get_simplefin_sync_settings(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn set_simplefin_sync_settings(
    state: tauri::State<'_, AppState>,
    settings: crate::sync_scheduler::SimpleFinSyncSettings,
) -> AppResult<()> {
    finsight_api::commands::simplefin::set_simplefin_sync_settings(&state.api, settings).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_simplefin_alerts(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<SimpleFinAlert>> {
    finsight_api::commands::simplefin::list_simplefin_alerts(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn acknowledge_simplefin_alert(
    state: tauri::State<'_, AppState>,
    alert_id: String,
) -> AppResult<()> {
    finsight_api::commands::simplefin::acknowledge_simplefin_alert(&state.api, alert_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_simplefin_transfer_suggestions(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<TransferSuggestionInfo>> {
    finsight_api::commands::simplefin::list_simplefin_transfer_suggestions(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn confirm_simplefin_transfer(
    state: tauri::State<'_, AppState>,
    transfer_id: String,
) -> AppResult<()> {
    finsight_api::commands::simplefin::confirm_simplefin_transfer(&state.api, transfer_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn reject_simplefin_transfer(
    state: tauri::State<'_, AppState>,
    transfer_id: String,
) -> AppResult<()> {
    finsight_api::commands::simplefin::reject_simplefin_transfer(&state.api, transfer_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_import_review_candidates(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<ImportCandidateWithMatches>> {
    finsight_api::commands::simplefin::list_import_review_candidates(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn accept_import_candidate_match(
    state: tauri::State<'_, AppState>,
    candidate_id: String,
    transaction_id: String,
) -> AppResult<()> {
    finsight_api::commands::simplefin::accept_import_candidate_match(
        &state.api,
        candidate_id,
        transaction_id,
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn create_import_candidate_transaction(
    state: tauri::State<'_, AppState>,
    candidate_id: String,
) -> AppResult<String> {
    finsight_api::commands::simplefin::create_import_candidate_transaction(
        &state.api,
        candidate_id,
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn dismiss_import_candidate(
    state: tauri::State<'_, AppState>,
    candidate_id: String,
) -> AppResult<()> {
    finsight_api::commands::simplefin::dismiss_import_candidate(&state.api, candidate_id).await
}
